DEV: to-markdown prosemirror HTML parsing

renato opened the pull request:
Replaces the custom ~1000-line `to-markdown` converter with ProseMirror's schema-based parsing and serialization pipeline, reusing the same extensions that power the rich editor. The old converter duplicated logic the extensions already implement (image formatting, mention/hashtag serialization, table handling, list nesting, etc.) — a single pipeline means one set of rules, fewer divergence bugs, and plugin extensions work in both directions automatically.

### Core change

- `toMarkdown()` is now async and lazy-loads ProseMirror on first use.
- Cooked→markdown serialization runs through the registered rich-editor extensions, so both conversion directions share one source of truth.

### QuoteState and quote callers

Making `toMarkdown()` async changed the quote flow. `QuoteState` now stores the selection's HTML and exposes an async `markdown()`; `buffer` is the plain-text selection. Every caller that needs markdown was updated to await `markdown()`:

- Topic controller (`selectText`, `replyToPost` including EmbedMode, `replyAsNewTopic`, `buildQuoteMarkdown`).
- Nested controller — forwards the full selection state via the new `copyFrom`.
- Fast-edit modal and `computeSupportsFastEdit`.
- discourse-ai post-helper menu — now sends markdown (not plain text) to the AI endpoints.

`markdown()` snapshots `opts` before awaiting so a concurrent selection change can't mis-pair the result.

### Plugin API

- Removed plugin-specific `addTagDecorateCallback`/`addTextDecorateCallback` usage from local-dates and spoiler-alert (the old APIs are kept as deprecated no-ops). local-dates now uses a `transformParsedHTML` rich-editor extension hook.

### Extractions and sharing

- Word paste handling moved into a dedicated `word-paste.js` extension.
- Quote-selection list-structure preservation moved into `preserve-list-structure.js`.
- `normalizeTable` is shared between the paste plugin and the serializer for consistent table output across editor modes.

### Quoting fix

- `selectedHTML` drops the empty trailing block a triple-click leaves behind. A triple-click extends the selection's end to the start of the following block, so `cloneContents()` would otherwise clone an empty `<blockquote>` that serializes to a stray `> ` in the quote.

renato committed:
DEV: to-markdown prosemirror HTML parsing

SamSaffron committed:
DEV: Improve quote transfer and list handling in rich editor

Preserve list structure when copying/pasting selections that span
multiple lists of different types (ordered/unordered). Annotates
orphan list items with their original list metadata before wrapping.

Key changes:
- Add list preservation logic to selectedText() utility function
- Support Word document list paste transformation
- Improve table serialization with validation and text fallback
- Add lightbox image parsing to preserve original upload URLs
- Handle mentions with user status icons in parseDOM
- Skip data URI images without upload permissions
- Add grid, wrap, and override-drag-ghost extensions
- Move math extension tests to plugin directory

SamSaffron committed:
DEV: Use ProseMirror for to-markdown conversion

This commit completes the migration of toMarkdown() to use
ProseMirror-based HTML parsing instead of the legacy custom parser.

Key changes:
- Remove ~800 lines of legacy Tag/Element parsing classes from
  to-markdown.js that are now handled by ProseMirror extensions
- Add deprecation warnings to addTagDecorateCallback,
  addBlockDecorateCallback, and addTextDecorateCallback
- Add ensureDefaultExtensions() to load extensions synchronously
  when toMarkdown() is called outside the editor context
- Add transformParsedHTML hook support for extensions to pre-process
  DOM before ProseMirror parsing

Plugin updates:
- discourse-local-dates: Move date range handling to rich-editor
  extension using new transformParsedHTML hook to wrap adjacent
  date spans in a container element for proper parsing
- spoiler-alert: Remove legacy to-markdown callbacks (now handled
  by rich-editor-extension), fix text selection in cooked content

Updates related tests to work with new ProseMirror-based output.

SamSaffron committed:
DEV: Extract list preservation logic into dedicated modules

Refactor list structure handling for selection/quote operations:

- Extract preserve-list-structure.js from utilities.js
- Create shared list-utils.js for common list utilities
- Add explicit state machine for orphan LI run collection
- Track default extensions registration to prevent duplicates
- Simplify deprecated to-markdown callback API to no-ops
- Add <del> tag support to strikethrough extension
- Add early exit optimizations to table validation
- Add unit tests for preserve-list-structure module

renato committed:
DEV: async load the deps on to-markdown

renato committed:
DEV: attempt to fix tests

renato committed:
DEV: attempt to fix tests #2

renato committed:
DEV: normalize table structure (for parity with HTML paste on rich mode)

renato committed:
DEV: test regression

renato committed:
DEV: remove validateTable logic, normalization should cover it

renato committed:
DEV: table testing

renato committed:
Merge branch 'main' into use_pm_on_to_markdown

renato committed:
DEV: fix test?

renato committed:
DEV: fix test??

megothss commented on frontend/discourse/tests/unit/lib/selection/preserve-list-structure-test.js:
```diff
@@ -0,0 +1,97 @@
+import { module, test } from "qunit";
+import { collectContiguousOrphanRuns } from "discourse/lib/selection/preserve-list-structure";
+
+module("Unit | Lib | selection | preserve-list-structure", function () {
```
Should we add more tests? 
List structures can get complicated with nested lists, test partial selections, etc.

megothss commented on frontend/discourse/app/static/prosemirror/extensions/link.js:
```diff
@@ -22,8 +22,12 @@ const extension = {
-          tag: "a[href]",
+          tag: "a[href]:not(.lightbox)",
           getAttrs(dom) {
+            if (dom.textContent === dom.getAttribute("href")) {
```
```suggestion
            if (dom.textContent.trim() === dom.getAttribute("href")) {
```

megothss commented on frontend/discourse/app/lib/to-markdown.js:
```diff
@@ -1,967 +1,82 @@
+    const serializer = new Serializer(extensions, pluginParams);
+
+    const processedHtml = transformWordListsHtml(html);
+    const element = new DOMParser().parseFromString(processedHtml, "text/html");
```
`DOMParser().parseFromString()` returns a document. Can we rename the variable because `element` is misleading?

megothss commented on frontend/discourse/app/static/prosemirror/extensions/word-paste.js:
```diff
@@ -0,0 +1,241 @@
+    if (/^[a-zA-Z][.)]/.test(markerText)) {
+      return { isOrdered: true, startNumber: 1 };
+    }
+    if (/^[ivxIVX]+[.)]/.test(markerText)) {
```
Same as above? LCDM?

renato committed:
DEV: address review feedback on quote state and topic controller

renato committed:
DEV: address more review comments

renato committed:
FIX: preserve full:true in rich editor quote serialization

renato committed:
DEV: add tests

megothss commented on frontend/discourse/app/components/cook-text.gjs:
```diff
@@ -19,9 +20,12 @@ export default class CookText extends Component {
+    waitForPromise(
+      cook(this.args.rawText).then((cooked) => {
+        this.cooked = cooked;
+      })
```
In theory this can throw if you the component is destroyed before the promise is fullfilled.


```suggestion
      cook(this.args.rawText).then((cooked) => {
        if (!this.isDestroying && !this.isDestroyed) {
          this.cooked = cooked;
        }
      })
```

renato committed:
Merge branch 'main' into use_pm_on_to_markdown

renato committed:
DEV: address review

renato committed:
DEV: remove throw from toMarkdown

megothss commented on frontend/discourse/app/lib/textarea-text-manipulation.js:
```diff
@@ -489,7 +489,9 @@ export default class TextareaTextManipulation {
 
     if (canPasteHtml && !handled) {
-      let markdown = toMarkdown(html);
+      e.preventDefault();
```
Now that `e.preventDefault()` is before the `await`, if `toMarkdown` returns `""` and there's `plainText`, the paste gets swallowed silently (native paste is already blocked, but the length check skips the insert). On `main` this wasn't an issue since `e.preventDefault()` was only called when `handled === true`:

https://github.com/discourse/discourse/blob/375581ed9875407d4cbeda8cec86c96567955c90/frontend/discourse/app/lib/textarea-text-manipulation.js#L512-L513

Should we fall back to inserting `plainText` in that case?

```js
if (!plainText || plainText.length < markdown.length) {
  // ... existing insert logic ...
} else if (plainText) {
  this.eventPrefix
    ? this.appEvents.trigger(`${this.eventPrefix}:insert-text`, plainText)
    : this.insertText(plainText);
  handled = true;
}
```

megothss commented on frontend/discourse/app/lib/to-markdown.js:
```diff
@@ -1,968 +1,92 @@
+    `${name} is deprecated. Use api.registerRichEditorExtension() instead.`,
+    {
+      id: `discourse.to-markdown.${name}`,
+      since: "v2026.4",
```
Nit: `since` should be `"2026.4.0"` (no `v` prefix, three-part version matching `lib/version.rb`).

megothss commented on frontend/discourse/app/lib/to-markdown.js:
```diff
@@ -1,968 +1,92 @@
+export function addTextDecorateCallback() {
+  deprecationWarning("addTextDecorateCallback");
+}
+export function clearTagDecorateCallbacks() {}
```
The `add*` counterparts warn on deprecation, but the `clear*` functions are silent no-ops. Is that intentional? External plugins calling these get no feedback at all.

megothss commented on frontend/discourse/app/controllers/topic.js:
```diff
@@ -535,63 +535,65 @@ export default class TopicController extends Controller {
-        draftSequence: post.get("topic.draft_sequence"),
-        draftKey: post.get("topic.draft_key"),
-      };
+    if (this.isDestroying || this.isDestroyed) {
```
Nit: Ember controllers are singletons that don't get destroyed during route transitions, so this guard will never fire in practice.

megothss commented on frontend/discourse/app/static/prosemirror/extensions/quote.js:
```diff
@@ -104,14 +104,15 @@ const extension = {
+      const { username, displayName, postNumber, topicId, full } = node.attrs;
       const postNumberParam = postNumber ? `, post:${postNumber}` : "";
       const topicIdParam = topicId ? `, topic:${topicId}` : "";
+      const fullParam = full === "true" ? `, full:true` : "";
```
The serializer correctly uses strict comparison here (`=== "true"`), but `toDOM` uses `full ? "true" : "false"` which is truthy-based:

https://github.com/discourse/discourse/blob/d272380e134df41c54ff0f96d2c9765ba957972b/frontend/discourse/app/static/prosemirror/extensions/quote.js#L43

This can flip the attribute on a copy-paste-within-editor round-trip:

1. From markdown parse, `full` is `null` (no `full:true` in bbcode)
2. `toDOM` renders `data-full="false"` (because `null` is falsy)
3. Copy-paste triggers `parseDOM`, `dom.getAttribute("data-full")` returns string `"false"`
4. `toDOM` sees string `"false"` (truthy), renders `data-full="true"`
5. Serializer sees `"true"`, emits `full:true`

Pre-existing in `toDOM`, but now more impactful since this PR adds `full` to the serializer. Maybe change L43 to `full === "true" ? "true" : "false"`?

discourse-patch-triage[bot] committed:
Merge remote-tracking branch 'origin/main' into use_pm_on_to_markdown

# Conflicts:
#	frontend/discourse/app/controllers/topic.js
#	frontend/discourse/app/lib/to-markdown.js
#	plugins/discourse-calendar/test/javascripts/integration/components/dates-test.gjs
#	plugins/discourse-local-dates/test/javascripts/acceptance/local-dates-quoting-test.js
#	plugins/spoiler-alert/test/javascripts/unit/lib/to-markdown-test.js

SamSaffron commented:
@patch-triage can you rebase this and clean up anything that needs cleaning up.

SamSaffron commented:
@patch-triage try rebasing again... leave no unresolved markers ...

janzenisaac commented:
@patch-triage, rebase this.

discourse-patch-triage[bot] committed:
Merge remote-tracking branch 'origin/main' into use_pm_on_to_markdown

# Conflicts:
#	frontend/discourse/app/components/post-text-selection-toolbar.gjs
#	frontend/discourse/app/controllers/topic.js
#	frontend/discourse/app/lib/composer/rich-editor-extensions.js
#	frontend/discourse/app/static/prosemirror/extensions/register-default.js

janzenisaac commented:
@patch-triage fix the linting issues.

discourse-patch-triage[bot] committed:
FIX: Address to-markdown review followups

SamSaffron commented:
@patch-triage eslint and prettier ain't linting... fix that ...

discourse-patch-triage[bot] committed:
DEV: Fix to-markdown lint failures

SamSaffron commented:
@patch-triage perform a review on this code isolate any risks or concerns top 5 things we should improve 

do not fix anything yet

SamSaffron commented:
@patch-triage Write red tests for the top 3 issues and then fix them

discourse-patch-triage[bot] committed:
DEV: Add to-markdown regression coverage

renato committed:
FIX: Update QuoteState callers for the new buffer/markdown contract

PR #35563 changed `quoteState.buffer` from the markdown form of the
selection to plaintext, with markdown moved behind `await
quoteState.markdown()`. Several callers still expected `buffer` to carry
markdown, resulting in plaintext-shaped quotes in flows that need
markdown.

Caller fixes:

* `controllers/nested.js`: use the new `copyFrom` helper so the topic
  controller's QuoteState receives the source HTML and `markdown()` can
  do the conversion (was dropping `selectedHtml`/`cookedHtml`).
* `controllers/topic.js`: EmbedMode branch of `replyToPost` now awaits
  `quoteState.markdown()` like the non-EmbedMode branch.
* `components/post-text-selection.gjs`: FastEdit modal and
  `computeSupportsFastEdit` use the markdown form for both the raw
  uniqueness check and the modal `initialValue`.
* `plugins/discourse-ai/.../ai-post-helper-trigger.gjs`: snapshot
  building moved from eager field init to the action handler so it can
  await `markdown()`; `markdownPromise`, `postId`, and `selectedRange`
  are captured synchronously before any await so a concurrent selection
  change cannot poison the menu data.

QuoteState changes:

* `markdown()` snapshots `opts` before the await so a concurrent
  `selected()` call cannot pair the result with a later selection's
  attribution.
* `get buffer()` lazily derives plaintext from `selectedHtml` using a
  `<template>` (inert DocumentFragment, no resource fetches). The
  `plainText` slot in `selected()` is now unused; the slot is kept to
  preserve the legacy 5-arg shape for back-compat.
* Add `copyFrom(other)` for transferring full state between instances
  (used by the nested controller).
* Bump `since` on the `addTagDecorateCallback` family deprecations to
  the release version.
* Add `quote-state-test.js` covering opts race, inflight survival across
  `clear()`, derivation, `copyFrom`, and full-flag detection.

renato committed:
Merge remote-tracking branch 'origin/main' into use_pm_on_to_markdown

# Conflicts:
#	frontend/discourse/app/ui-kit/d-cook-text.gjs
#	spec/system/script_encoding_spec.rb

renato committed:
FIX: Drop the empty trailing block a triple-click adds to a quote

Triple-clicking a paragraph extends the browser selection's end to the very
start of the following block (endOffset 0). `selectedHTML`'s
`range.cloneContents()` then materialises that block as an empty husk — e.g.
`<blockquote>\n<p></p></blockquote>` for the "> If you need help…" line in the
new-user welcome message — which the markdown serializer renders as a stray
`> ` in the quote.

`selectedHTML` now pulls the range end back past that husk before cloning:
when `endOffset === 0` it walks the boundary up the ancestor chain while the
node is the first meaningful child of its parent (skipping the whitespace text
nodes cooked HTML places between blocks) and `setEndBefore`s it, so the husk is
never cloned. The walk stops at the first node with real preceding content, so
a precise selection ending right after an image keeps `<p><img></p>`.
